### PR TITLE
topk overhaul and argtopk

### LINKED ISSUE #2965: Corner cases for topk of array
Kind of silly, but I run into this when I try to define topk(k, x, axis=None).
Test:
```python
import numpy as np
import dask.array as da
x = da.from_array(np.array([0]), chunks=1)
da.topk(x, 0)
```
Result:
```python
Traceback (most recent call last):
  File "test_topk2.py", line 4, in <module>
    da.topk(x, 0)
  File "<python dir>/site-packages/dask/array/routines.py", line 790, in topk
    if x.ndim != 1:
```

Tests:
```python
import numpy as np
import dask.array as da
cases = [(0, 0, 0), (0, 1, 0), (0, 1, 1), ([0], 1, 0), ([0], 1, 1)]

for a, b, c in cases:
    x = da.from_array(np.array(a), chunks=b)
    try:
        da.topk(x, c)
    except Exception as err:
        print("[ case: (%s, %s, %s) ]" % (a, b, c))
        print('fail', err)
```
Result:
```
[ case: (0, 0, 0) ]
fail 'int' object has no attribute 'ndim'
[ case: (0, 1, 0) ]
fail 'int' object has no attribute 'ndim'
[ case: (0, 1, 1) ]
fail 'int' object has no attribute 'ndim'
[ case: ([0], 1, 0) ]
fail 'int' object has no attribute 'ndim'
[ case: ([0], 1, 1) ]
fail 'int' object has no attribute 'ndim'
```

### LINKED ISSUE #3395: Recursive topk
I have an array with shape (500000, ) and chunks=2500.
I need to pick the top 5000 elements out of it.

The current topk implementation is very bad for it, as:
1. it applies one kernel per chunk of the original input
2. concatenates all the outputs of the first pass together into a single chunk
3. performs a single, final sort

As my k is greater than the chunk size, effectively step 2 would create a single-chunk array or 500000 points.
The workaround would be to rechunk the array ahead of topk to larger chunks - which is often not the easiest thing to do when you're implementing a general-purpose library call.

As a solution I suggest to add a recursive=False option to both the method and the function; see prototype below:

https://gist.github.com/crusaderky/35d355d8124141aeea82213e0c7bccde

### LINKED ISSUE #3396: Risk attribution: a[argsort(b)[:k]]
# Abstract problem
I have ``y = f(x1, x2, ...)``, where ``x1, x2...`` are 1-dimensional arrays with the same shape, and ``f`` is a generic elementwise, embarassingly parallel function (e.g. ``sum()``). As a consequence, ``y`` has the same shape as the inputs.
I need to find:
1. the top k elements of ``y``
2. the elements of ``x1, x2,...`` which produced the top k elements of ``y``

# Real life use case
``x1, x2...`` are the values of financial assets under a Monte Carlo risk simulation, one point per scenario, and y is the value of the portfolio containing them. I need to find
1. the 99% worst value of y, e.g. where the portfolio loses the most money (_Value at Risk_, or _VaR_)
2. the value of each individual asset building the portfolio where the portfolio is the worst 99% (_risk attribution_) 

In qualitative terms, (2) answers the question: _When you're on the brink of bankrupting, what is causing you to do so? Which of the products that you own are a time bomb that you should sell NOW if you fancy surviving a disaster, e.g. a new global market crisis?_

# Solution
The fist point is readily solvable with topk: 
```
k = int(round(y.size * .01)
-(-y).topk(k)[-1]
```
The second point is not possible in dask to my understanding; in numpy it would be: ``x[argsort(y)[k]]``

There's two things that are missing to make it happen in dask:
1. an ``argtopk`` function, which returns the indexes of the top k elements of an array
2. the ability to slice ``a[b]``, where b is a dask array of integers.

With these, the dask solution to the problem would be ``x[(-y).argtopk(k)[-1]]``

Working on a POC - stay tuned...

### PR DESCRIPTION
**Pull request by crusaderky**
topk changes:
+ change API from topk(k, a) to topk(a, k) (closes #2965)
+ add support for multiple axis
+ add support for recursive aggregation (closes #3395)
+ avoid needless internal sorting in intermediate passages
+ new function argtopk (partially resolves #3396)

### CODE REVIEW

File: dask/array/chunk.py [hunk 1]
```diff
@@ -176,16 +176,50 @@ def trim(x, axes=None):
...
+    # a = a[-k:] if k>0 else a[:-k], on arbitrary axis
+    a = a[[
+        (slice(-k, None) if k > 0 else slice(None, -k))
+        if i == axis else slice(None)
+        for i in range(a.ndim)
+    ]]
+    return a
+
+
+def topk_postproc(a, k, axis):
```
Comment (jcrist) [inline dask/array/chunk.py]: These names show up in the graphs, and are only marginally shorter than the non-abbreviated versions. I'd prefer `topk_postprocess` and `argtopk_preprocess`.

File: dask/array/core.py [hunk 2]
```diff
@@ -1401,12 +1401,19 @@ def reshape(self, *shape):
...
         """The top k elements of an array.
 
         See ``da.topk`` for docstring"""
-        from .routines import topk
-        return topk(k, self)
+        from .reductions import topk
+        return topk(self, k, axis=axis, split_every=split_every)
+
+    def argtopk(self, k, axis=-1, split_every=None):
+        """The indexes of the top k elements of an array.
```
Comment (jcrist) [inline dask/array/core.py]: `indexes -> indices`

File: dask/array/reductions.py [hunk 3]
```diff
@@ -722,3 +723,70 @@ def validate_axis(ndim, axis):
...
+    >>> import dask.array as da
+    >>> x = np.array([5, 1, 3, 6])
+    >>> d = da.from_array(x, chunks=2)
+    >>> d.topk(2).compute()
+    array([6, 5])
+    >>> d.topk(-2).compute()
+    array([1, 3])
+    """
+    if isinstance(a, int) and isinstance(k, Array):
+        warnings.warn("DeprecationWarning: topk(k, x) has been replaced with topk(a, k)")
```
Comment (jcrist) [inline dask/array/reductions.py]: Should be `topk(k, a) has bee been replaced with topk(a, k)`

File: dask/array/reductions.py [hunk 4]
```diff
@@ -722,3 +723,70 @@ def validate_axis(ndim, axis):
...
+    chunks = list(res.chunks)
+    chunks[axis] = (abs(k), )
+    res = Array(res.dask, res.name, chunks, res.dtype)
+
+    # Sort result internally
+    return res.map_blocks(chunk.topk_postproc, k=k, axis=axis, dtype=a.dtype)
+
+
+def argtopk(a, k, axis=-1, split_every=None):
+    """Extract the indexes of the k largest elements from a on the given axis,
```
Comment (jcrist) [inline dask/array/reductions.py]: `indexes -> indices` throughout this docstring.

File: dask/array/reductions.py [hunk 5]
```diff
@@ -722,3 +723,70 @@ def validate_axis(ndim, axis):
...
+    >>> d = da.from_array(x, chunks=2)
+    >>> d.topk(2).compute()
+    array([6, 5])
+    >>> d.topk(-2).compute()
+    array([1, 3])
+    """
+    if isinstance(a, int) and isinstance(k, Array):
+        warnings.warn("DeprecationWarning: topk(k, x) has been replaced with topk(a, k)")
+        a, k = k, a
+
```
Comment (jcrist) [inline dask/array/reductions.py]: Since the rest of this code assumes axis >= 0, you should use `validate_axis` here.

```python
axis = validate_axis(a.ndim, axis)
```

This will handle negative axis, and throw a nice error for invalid values.

File: dask/array/reductions.py [hunk 6]
```diff
@@ -722,3 +723,70 @@ def validate_axis(ndim, axis):
...
+    >>> x = np.array([5, 1, 3, 6])
+    >>> d = da.from_array(x, chunks=2)
+    >>> d.argtopk(2).compute()
+    array([3, 0])
+    >>> d.argtopk(-2).compute()
+    array([1, 2])
+    """
+    # Convert a to a recarray that contains its index
+    if axis < 0:
+        axis += a.ndim
```
Comment (jcrist) [inline dask/array/reductions.py]: I recommend using `validate_axis` here instead.

File: dask/array/reductions.py [hunk 7]
```diff
@@ -722,3 +723,70 @@ def validate_axis(ndim, axis):
...
+    idx = arange(a.shape[axis], chunks=a.chunks[axis])
+    idx = idx[tuple(slice(None) if i == axis else np.newaxis for i in range(a.ndim))]
+    a_rec = a.map_blocks(chunk.argtopk_preproc, idx, dtype=[('values', a.dtype), ('idx', int)])
```
Comment (jcrist) [inline dask/array/reductions.py]: On windows `int` will result in `int32`. It's not clear to me if that's bad or not, just thought I'd bring it up.

File: dask/array/tests/test_reductions.py [hunk 8]
```diff
@@ -422,3 +422,45 @@ def test_array_cumreduction_out(func):
...
+def test_topk_argtopk():
+    # Test data
+    k = 5
+    a = da.random.random(1000, chunks=250)
+    b = da.random.random((10, 20, 30), chunks=(4, 8, 8))
+    c = da.from_array(np.array([(1, 'Hello'), (2, 'World')], dtype=[('foo', int), ('bar', '<U5')]),
+                      chunks=1)
+
+    # Support for deprecated API for topk
+    np.testing.assert_array_equal(da.topk(a, 5), da.topk(5, a))
```
Comment (jcrist) [inline dask/array/tests/test_reductions.py]: We should check that the warning is thrown.

```python
with pytest.warns(UserWarning):
    assert_eq(da.topk(a, 5), da.topk(5, a))
```

File: dask/array/tests/test_reductions.py [hunk 9]
```diff
@@ -422,3 +422,45 @@ def test_array_cumreduction_out(func):
...
+
+    # As Array methods
+    a.topk(5)
+    np.testing.assert_array_equal(da.topk(a, 5), da.topk(5, a))
```
Comment (jcrist) [inline dask/array/tests/test_reductions.py]: In dask we make use of `assert_eq` to test equality. This checks not only values, but also the metadata on the dask arrays, which helps ensure correctness. You should be able to replace all your calls to `assert_array_equal` with this.

File: dask/array/tests/test_reductions.py [hunk 10]
```diff
@@ -422,3 +422,45 @@ def test_array_cumreduction_out(func):
...
+                                          daskf(b, k, axis=1, split_every=se))
+            np.testing.assert_array_equal(npf(b, axis=2)[:, :, -k:][:, :, ::-1],
+                                          daskf(b, k, axis=2, split_every=se))
+
+            np.testing.assert_array_equal(npf(b, axis=0)[:k, :, :],
+                                          daskf(b, -k, axis=0, split_every=se))
+            np.testing.assert_array_equal(npf(b, axis=1)[:, :k, :],
+                                          daskf(b, -k, axis=1, split_every=se))
+            np.testing.assert_array_equal(npf(b, axis=2)[:, :, :k],
+                                          daskf(b, -k, axis=2, split_every=se))
```
Comment (jcrist) [inline dask/array/tests/test_reductions.py]: Nice tests. Can I also ask that you add one with a negative axis?

File: dask/array/reductions.py (hunk 3 above)
Comment (crusaderky) [reply dask/array/reductions.py]: No, the previous API really was topk(k, x). I renamed the array to `a` to be coherent with all other reduction functions.

Comment (jcrist) [reply dask/array/reductions.py]: Yes, but since the argument was positional, as a reader all I really care about is the swapping of the parameter order (array and `k`). As it currently reads it's not immediately clear that the order of the array and `k` parameter have swapped.

**Comment by crusaderky** [comment]
@jcrist I integrated your code review and overhauled the unit test.
Travis fails on something completely unrelated to what I touched - could you look into it please?

**Comment by jcrist** [comment]
Travis failure was due to an OOM error (if you scroll up you can see the originating exception). Restarted build and everything passes. Thanks for the contribution, merging.

File: docs/source/changelog.rst [hunk 11]
```diff
@@ -12,6 +12,12 @@ Array
 - Add ``piecewise`` for Dask Arrays (:pr:`3350`) `John A Kirkham`_
 - Fix handling of ``nan`` in ``broadcast_shapes`` (:pr:`3356`) `John A Kirkham`_
 - Add ``isin`` for dask arrays (:pr:`3363`). `Stephan Hoyer`_
+- Overhauled ``topk`` for Dask Arrays: faster algorithm, particularly for large k's; added support
+  for multiple axes, recursive aggregation, and an option to pick the bottom k elements instead.
+  (:pr:`3395`) `Guido Imperiale`_
+- The ``topk`` API has changed from topk(k, array) to the more conventional topk(array, k).
+  The legacy API still works but is now deprecated. (:pr:`2965`) `Guido Imperiale`_
```
Comment (jakirkham) [inline docs/source/changelog.rst]: Was this suppose to be `2965` or was it meant to be `3395` as well?
